doc/architecture/decisions/0015-container-api.md

```diff
@@ -0,0 +1,103 @@
+as a special case, since this does not allow for app composability (which we think is a crucial
+concept for the Holochain app ecosystem). 
+
+Instead, in the context of the Rust iteration of Holochain, we have been following the notion of
+providing a *Container* as a relevant concept (and implemention) for the deployment of Holochain apps:
+we provide a Container (i.e. [HoloSqape](https://github.com/holochain/holosqape)) for each supported platform (Linux, MacOS, Windows, iOS, Android)
+that gets installed on a host machine. Holochain apps get installed into the container.
+The Container:  
+ * manages installed hApps (Holochain Apps),
+ * manages agent identities (keys) ,
+ * should also enable hApps to call functions of other hApps - what we call *bridging*,
+ * has to implement access permissions to installed hApps.
+
+So far, the interface our Container implementation provides was growing organically
+in [container.h](https://github.com/holochain/holosqape/blob/master/bindings/container.h).
+
+With upcoming alternative container implementations (based on [Holochain-nodejs](https://github.com/holochain/holochain-nodejs)
+or a custom one for HoloPorts) we should drive the process of building this Container API
+consciously and with more coherence and visibility amongst our separate dev teams.
+
+We need a protocol for communication with a Holochain container and a specification of what upcoming
+containers have to implement, so that apps and UIs can be build against a standardised
+
+## Decision
+
+We establish the **Container as a fundamental module of the Holochain architecture/stack**
+by specifying its **API**, that can be assumed by UI components, Holochain apps (i.e. zome
+functions in the case of bridging)
+and remote processes alike to be implemented by the context a Holochain app is executed in.
+
+
+Fundamental to this API is user roles and permissions.
+Users will be able to use different subsets of the Container's API depending on their specific permissions. 
+
+### Generic API
+
+Every user (of the Container API, i.e. QML root widgets, admin UI in the case of HoloSqape, 
+and network services built on top of the Container) 
+will have these functions available, though functions can return with a permission denied
+error in case an app or capability was given that the current user is not allowed to use.
+
+* `call(app_hash, capability_name, function_name, parameters) -> {results}`
+* `connect(app_hash, capability_name, signal_name, callback_function)`
+* `installed_apps() -> [hash]`
+* `capabilities_of(app_hash) -> [String]`
+* `functions_of(app_hash, capability_name) -> [String]`
+* `request_access(app_hash, capability)`
+* `can_access?(app_hash, capability) -> bool`
+
+
+### Admin
+
+Only the admin user can manage apps and permissions:
+
+* `install_app(Dna) -> hash`
+* `uninstall_app(hash)`
+* `start_app(hash)`
+* `stop_app(hash)`
+* `promote_to_admin(user_id)`
```

Comment:
What is a "user_id" in this context?

Reply:
The idea is that a "user" is who ever is using the container, i.e. not a human, but rather a particular UI, or a particular connection over the websocket connector.  This name will probably have to change as it's confusing.

Reply:
Right... I'd recommend "client" then, seems fitting

Reply:
I am also thinking out though... where exactly does this ID come from? As in... how does a client get assigned an ID, and where is that ID "registered". This seems an important question to clarify just a little bit more an important aspect of this